staging/src/k8s.io/apimachinery/pkg/runtime/helper.go

```diff
@@ -236,6 +236,10 @@ func (e WithVersionEncoder) Encode(obj Object, stream io.Writer) error {
 			gvk = preferredGVK
 		}
 	}
+	if gvk == oldGVK {
+		// gvk already set as intended, skip injecting it.
+		return e.Encoder.Encode(obj, stream)
+	}
 	kind.SetGroupVersionKind(gvk)
```

Comment:
would it be more readable/robust to do:

```go
if gvk != oldGVK {
  kind.SetGroupVersionKind(gvk)
  defer kind.SetGroupVersionKind(oldGVK)
}
return e.Encoder.Encode(obj, stream)
```

Reply:
It's not just a shortcut anymore, but the resulting code is better - I'll change to that.

Reply:
Done and PR description + commit message updated.